public/app/plugins/datasource/graphite/datasource_integration.test.ts

```diff
@@ -0,0 +1,154 @@
+import { GraphiteDatasource } from './datasource';
+import { TemplateSrv } from 'app/features/templating/template_srv';
+import { BackendSrv } from 'app/core/services/backend_srv';
+import { of } from 'rxjs';
+import { ContextSrv, User } from '../../../core/services/context_srv';
+import { setBackendSrv } from '@grafana/runtime';
+
+jest.mock('@grafana/runtime', () => ({
+  ...(jest.requireActual('@grafana/runtime') as unknown as object),
+}));
+
+interface Context {
+  templateSrv: TemplateSrv;
+  ds: GraphiteDatasource;
+}
+
+describe('graphiteDatasource integration with backendSrv and fetch', () => {
+  let ctx = {} as Context;
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+
+    const instanceSettings = {
+      url: '/api/datasources/proxy/1',
+      name: 'graphiteProd',
+      jsonData: {
+        rollupIndicatorEnabled: true,
+      },
+    };
+    const templateSrv = new TemplateSrv();
+    const ds = new GraphiteDatasource(instanceSettings, templateSrv);
+    ctx = { templateSrv, ds };
+  });
+
+  describe('returns a list of functions', () => {
+    it('should return a list of functions with invalid JSON', async () => {
+      const INVALID_JSON =
+        '{"testFunction":{"name":"function","description":"description","module":"graphite.render.functions","group":"Transform","params":[{"name":"param","type":"intOrInf","required":true,"default":Infinity}]}}';
+
+      mockBackendSrv(INVALID_JSON);
+
+      const funcDefs = await ctx.ds.getFuncDefs();
+
+      expect(funcDefs).toEqual({
+        testFunction: {
+          category: 'Transform',
+          defaultParams: ['inf'],
+          description: 'description',
+          fake: true,
+          name: 'function',
+          params: [
+            {
+              multiple: false,
+              name: 'param',
+              optional: false,
+              options: undefined,
+              type: 'int_or_infinity',
+            },
+          ],
+        },
+      });
+    });
+
+    it('should return a list of functions with valid JSON', async () => {
+      const VALID_JSON =
+        '{"testFunction":{"name":"function","description":"description","module":"graphite.render.functions","group":"Transform","params":[{"name":"param","type":"intOrInf","required":true,"default":1e9999}]}}';
+
+      mockBackendSrv(VALID_JSON);
+
+      const funcDefs = await ctx.ds.getFuncDefs();
+
+      expect(funcDefs).toEqual({
+        testFunction: {
+          category: 'Transform',
+          defaultParams: ['inf'],
+          description: 'description',
+          fake: true,
+          name: 'function',
+          params: [
+            {
+              multiple: false,
+              name: 'param',
+              optional: false,
+              options: undefined,
+              type: 'int_or_infinity',
+            },
+          ],
+        },
+      });
+    });
+  });
+});
+
+function mockBackendSrv(data: string) {
```

Comment:
I think this might be too excessive. It should be enough to mock the fetch with the following:

```typescript
function mockBackendSrv(data: string) {
  const fetchMock = jest.fn().mockReturnValue(of({ data }));
  setBackendSrv({ fetch: fetchMock } as any);
}
```

Reply:
I hear you on it being excessive, but we wanted to test the fromFetch dependency in backendSrv, which would then call parseResponseBody from /core/utils/fetch.ts. By only mocking the fetch, we don't get the full integration test with parseResponseBody.

Reply:
Oh yes, of course 🤦  Why do we need to mock `parseRequestOptionsMock` though?

Reply:
That's true! we don't need ```parseRequestOptions```, since the options are already in the mocked fromFetch. I removed the mock for ```parseRequestOptions```.